pandas/core/internals/blocks.py

```diff
@@ -5393,518 +3469,117 @@ def _putmask_preserve(nv, n):
-        # are only null blocks, when same upcasting rules must be applied to
-        # null upcast classes.
-        if unit.is_na:
-            null_upcast_classes[upcast_cls].append(dtype)
-        else:
-            upcast_classes[upcast_cls].append(dtype)
-
-    if not upcast_classes:
-        upcast_classes = null_upcast_classes
-
-    # create the result
-    if 'object' in upcast_classes:
-        return np.dtype(np.object_), np.nan
-    elif 'bool' in upcast_classes:
-        if has_none_blocks:
-            return np.dtype(np.object_), np.nan
-        else:
-            return np.dtype(np.bool_), None
-    elif 'category' in upcast_classes:
-        return np.dtype(np.object_), np.nan
-    elif 'datetimetz' in upcast_classes:
-        dtype = upcast_classes['datetimetz']
-        return dtype[0], tslibs.iNaT
-    elif 'datetime' in upcast_classes:
-        return np.dtype('M8[ns]'), tslibs.iNaT
-    elif 'timedelta' in upcast_classes:
-        return np.dtype('m8[ns]'), tslibs.iNaT
-    else:  # pragma
-        g = np.find_common_type(upcast_classes, [])
-        if is_float_dtype(g):
-            return g, g.type(np.nan)
-        elif is_numeric_dtype(g):
-            if has_none_blocks:
-                return np.float64, np.nan
-            else:
-                return g, None
-
-    msg = "invalid dtype determination in get_concat_dtype"
-    raise AssertionError(msg)
-
-
-def concatenate_join_units(join_units, concat_axis, copy):
-    """
-    Concatenate values from several join units along selected axis.
-    """
-    if concat_axis == 0 and len(join_units) > 1:
-        # Concatenating join units along ax0 is handled in _merge_blocks.
-        raise AssertionError("Concatenating join units along axis0")
-
-    empty_dtype, upcasted_na = get_empty_dtype_and_na(join_units)
-
-    to_concat = [ju.get_reindexed_values(empty_dtype=empty_dtype,
-                                         upcasted_na=upcasted_na)
-                 for ju in join_units]
-
-    if len(to_concat) == 1:
-        # Only one block, nothing to concatenate.
-        concat_values = to_concat[0]
-        if copy:
-            if isinstance(concat_values, np.ndarray):
-                # non-reindexed (=not yet copied) arrays are made into a view
-                # in JoinUnit.get_reindexed_values
-                if concat_values.base is not None:
-                    concat_values = concat_values.copy()
+# Also needed by managers
+def _extend_blocks(result, blocks=None):
+    """ return a new extended blocks, givin the result """
+    from .managers import BlockManager
+    if blocks is None:
+        blocks = []
+    if isinstance(result, list):
+        for r in result:
+            if isinstance(r, list):
+                blocks.extend(r)
             else:
-                concat_values = concat_values.copy()
+                blocks.append(r)
+    elif isinstance(result, BlockManager):
```

Comment:
This is the only place where internals.blocks depends on either `internals.concat` or `internals.managers`.  DAG FTW